create_parameter_weights.py

```diff
@@ -1,21 +1,88 @@
 # Standard library
 import os
+import subprocess
 from argparse import ArgumentParser
 
 # Third-party
 import numpy as np
 import torch
+import torch.distributed as dist
+from torch.utils.data.distributed import DistributedSampler
 from tqdm import tqdm
 
 # First-party
 from neural_lam import constants
 from neural_lam.weather_dataset import WeatherDataset
 
 
-def main():
-    """
-    Pre-compute parameter weights to be used in loss function
-    """
+def get_rank():
+    """Get the rank of the current process in the distributed group."""
+    if "SLURM_PROCID" in os.environ:
+        return int(os.environ["SLURM_PROCID"])
+    return 0
+
+
+def get_world_size():
+    """Get the number of processes in the distributed group."""
+    if "SLURM_NTASKS" in os.environ:
+        return int(os.environ["SLURM_NTASKS"])
+    return 1
+
+
+def setup(rank, world_size):  # pylint: disable=redefined-outer-name
+    """Initialize the distributed group."""
+    try:
+        master_node = (
+            subprocess.check_output(
+                "scontrol show hostnames $SLURM_JOB_NODELIST | head -n 1",
+                shell=True,
+            )
+            .strip()
+            .decode("utf-8")
+        )
+    except Exception as e:
+        print(f"Error getting master node IP: {e}")
+        raise
+    master_port = "12355"
+    os.environ["MASTER_ADDR"] = master_node
+    os.environ["MASTER_PORT"] = master_port
+    if torch.cuda.is_available():
+        dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    else:
+        dist.init_process_group("gloo", rank=rank, world_size=world_size)
+    print(
+        f"Initialized {dist.get_backend()} process group with "
+        f"world size "
+        f"{world_size}."
+    )
+
+
+def cleanup():
+    """Destroy the distributed group."""
+    dist.destroy_process_group()
+
+
+def adjust_dataset_size(ds, world_size, batch_size):
```

Comment:
Is my understanding correct that the way the dataset size is adjusted is by ignoring the last samples? While that does not change the statistics much for large datasets I think it needs to be clearly communicated when this script is used. It also means that the statistics will change somewhat depending on how many ranks you run this on right?

Reply:
True, I will resent another option that randomly draws saples from the dataset until all ranks are fully filled. At this point you might notice that I am not a computer engineer. :innocent: I will in any case add a comment about this and give the user a choise between speed and accuracy.

Reply:
Does that really solve the problem though? Then you would get different statistics depending on the random sampling.

What I am thinking more is that it would be nice if we guarantee that the true mean and std.-dev. of the training data is what we return here, no matter parallelization choices. One could for example distribute out as many samples as possible to fill out full batches on all ranks, run those batches in parallel, and then just run any samples left in batches only on rank 0 at the end. Then do the little bit of math to get combine this into the correct mean.

Alternatively, one could pad the dataset with duplicate samples to fill it up for all ranks, and then keep track of the padding samples to not count when reducing to the mean. Seems a bit trickier/messier to implement.

In the end I am not sure if it is worth it to put in the effort. Maybe it's enough to just document that this will drop some samples in the end of the dataset for the statistics computations.